lib/remoteproc/remoteproc_virtio.c

```diff
@@ -16,6 +16,140 @@
 #include <metal/utilities.h>
 #include <metal/alloc.h>
 
+static void rproc_virtio_delete_virtqueues(struct virtio_device *vdev)
+{
+	struct virtio_vring_info *vring_info;
+	unsigned int i;
+
+	if (vdev->vrings_info != NULL) {
+		for (i = 0; i < vdev->vrings_num; i++) {
+			vring_info = &vdev->vrings_info[i];
+			if (vring_info->vq != NULL) {
+				virtqueue_free(vring_info->vq);
+			}
+		}
+
+		metal_free_memory(vdev->vrings_info);
+	}
+}
+
+static int rproc_virtio_create_virtqueue(struct virtio_device *vdev,
+					 unsigned int flags,
+					 unsigned int i,
+					 const char *name,
+					 vq_callback callback)
+{
+	struct remoteproc_virtio *rpvdev;
+	struct fw_rsc_vdev_vring *vring_rsc;
+	struct fw_rsc_vdev *vdev_rsc;
+	struct remoteproc *rproc;
+	struct virtio_vring_info *vring_info;
+	struct vring_alloc_info *vring_alloc;
+	struct metal_io_region *io;
+	metal_phys_addr_t da;
+	size_t vringsize;
+	void *va;
+	int ret;
+	(void)flags;
+
+	/* Get remoteproc virtio device */
+	rpvdev = metal_container_of(vdev, struct remoteproc_virtio, vdev);
+
+	/* Get the remoteproc */
+	rproc = rpvdev->priv;
+
+	/* Get the rsc table */
+	vdev_rsc = rpvdev->vdev_rsc;
+	vring_rsc = &vdev_rsc->vring[i];
+
+	/*
+	 * Initialize the vring information according to the vring resource
+	 * table.
+	 */
+	da = vring_rsc->da;
+	vringsize = vring_size(vring_rsc->num, vring_rsc->align);
+	va = remoteproc_mmap(rproc, NULL, &da, vringsize, 0, &io);
```

Comment:
this can probably be kept in remoteproc_create_virtio

Reply:
Nice suggestions and Done.